compaction/compaction_manager.cc

```diff
@@ -297,9 +297,11 @@ future<> compaction_manager::run_custom_job(column_family* cf, sstables::compact
 
     auto job_ptr = std::make_unique<noncopyable_function<future<>()>>(std::move(job));
 
-    task->compaction_done = with_semaphore(_custom_job_sem, 1, [this, task, cf, &job = *job_ptr] () mutable {
+    task->compaction_done = with_semaphore(_custom_job_sem, 1, [this, task, cf, excl, &job = *job_ptr] () mutable {
         // take read lock for cf, so major compaction and resharding can't proceed in parallel.
-        return with_lock(_compaction_locks[cf].for_read(), [this, task, cf, &job] () mutable {
+        auto& lk = _compaction_locks[cf];
+        auto acquire_lock = excl ? lk.hold_write_lock() : lk.hold_read_lock();
+        return acquire_lock.then([this, task, cf, &job] (rwlock::holder h) mutable {
```

Comment:
h has to be forwarded to job() continuation, return job().then([h = std::move(h)]

Reply:
to avoid lock being released prematurely

Reply:
I dropped this patch in v2